File: src/main/java/me/syari/ss/discord/internal/utils/compress/ZlibDecompressor.java
package me.syari.ss.discord.internal.utils.compress;

import me.syari.ss.discord.api.utils.Compression;
import me.syari.ss.discord.internal.utils.IOUtil;
import me.syari.ss.discord.internal.utils.JDALogger;
import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.lang.ref.SoftReference;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.zip.DataFormatException;
import java.util.zip.Inflater;
import java.util.zip.InflaterOutputStream;

public class ZlibDecompressor implements Decompressor {
    private static final int Z_SYNC_FLUSH = 0x0000FFFF;

    private final int maxBufferSize;
    private final Inflater inflater = new Inflater();
    private ByteBuffer flushBuffer = null;
    private SoftReference<ByteArrayOutputStream> decompressBuffer = null;

    public ZlibDecompressor(int maxBufferSize) {
        this.maxBufferSize = maxBufferSize;
    }

    private SoftReference<ByteArrayOutputStream> newDecompressBuffer() {
        return new SoftReference<>(new ByteArrayOutputStream(Math.min(1024, maxBufferSize)));
    }

    private ByteArrayOutputStream getDecompressBuffer() {
        if (decompressBuffer == null)
            decompressBuffer = newDecompressBuffer();
        ByteArrayOutputStream buffer = decompressBuffer.get();
        if (buffer == null) {
            decompressBuffer = new SoftReference<>(buffer = new ByteArrayOutputStream(Math.min(1024, maxBufferSize)));
        }
        return buffer;
    }

    private boolean isFlush(byte @NotNull [] data) {
        if (data.length < 4) {
            return false;
        }
        int suffix = IOUtil.getIntBigEndian(data, data.length - 4);
        return suffix == Z_SYNC_FLUSH;
    }

    private void buffer(byte[] data) {
        if (flushBuffer == null)
            flushBuffer = ByteBuffer.allocate(data.length * 2);

        if (flushBuffer.capacity() < data.length + flushBuffer.position()) {
            flushBuffer.flip();
            flushBuffer = IOUtil.reallocate(flushBuffer, (flushBuffer.capacity() + data.length) * 2);
        }

        flushBuffer.put(data);
    }

    private Object lazy(byte[] data) {
        return JDALogger.getLazyString(() -> Arrays.toString(data));
    }

    @Override
    public Compression getType() {
        return Compression.ZLIB;
    }

    @Override
    public void reset() {
        inflater.reset();
    }

    @Override
    public void shutdown() {
        reset();
    }

    @Override
    public String decompress(byte[] data) throws DataFormatException {
        if (!isFlush(data)) {
            LOG.debug("Received incomplete data, writing to buffer. Length: {}", data.length);
            buffer(data);
            return null;
        } else if (flushBuffer != null) {
            LOG.debug("Received final part of incomplete data");
            buffer(data);
            byte[] arr = flushBuffer.array();
            data = new byte[flushBuffer.position()];
            System.arraycopy(arr, 0, data, 0, data.length);
            flushBuffer = null;
        }
        LOG.trace("Decompressing data {}", lazy(data));
        ByteArrayOutputStream buffer = getDecompressBuffer();
        try (InflaterOutputStream decompressor = new InflaterOutputStream(buffer, inflater)) {
            decompressor.write(data);
            return buffer.toString("UTF-8");
        } catch (IOException e) {
            throw (DataFormatException) new DataFormatException("Malformed").initCause(e);
        } finally {
            if (buffer.size() > maxBufferSize)
                decompressBuffer = newDecompressBuffer();
            else
                buffer.reset();
        }
    }
}

File: src/main/kotlin/me/syari/ss/discord/requests/ZlibDecompressor.kt
package me.syari.ss.discord.requests

import okhttp3.internal.and
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.lang.ref.SoftReference
import java.nio.ByteBuffer
import java.util.zip.DataFormatException
import java.util.zip.Inflater
import java.util.zip.InflaterOutputStream

internal object ZlibDecompressor {
    private const val maxBufferSize = 2048
    private const val Z_SYNC_FLUSH = 0x0000FFFF
    private val inflater = Inflater()
    private var flushBuffer: ByteBuffer? = null
    private var decompressBuffer: SoftReference<ByteArrayOutputStream>? = null

    private fun newDecompressBufferReferent(): ByteArrayOutputStream {
        return ByteArrayOutputStream(1024.coerceAtMost(maxBufferSize))
    }

    private fun newDecompressBuffer(): SoftReference<ByteArrayOutputStream> {
        return SoftReference(newDecompressBufferReferent())
    }

    private fun getDecompressBuffer(): ByteArrayOutputStream {
        val notNullDecompressBuffer = decompressBuffer ?: {
            newDecompressBuffer().apply {
                decompressBuffer = this
            }
        }.invoke()
        return notNullDecompressBuffer.get() ?: {
            newDecompressBufferReferent().apply {
                decompressBuffer = SoftReference(this)
            }
        }.invoke()
    }

    private fun getIntBigEndian(array: ByteArray, offset: Int): Int {
        return array[offset + 3] and 0xFF or (array[offset + 2] and 0xFF shl 8) or (array[offset + 1] and 0xFF shl 16) or (array[offset] and 0xFF shl 24)
    }

    private fun isFlush(data: ByteArray): Boolean {
        if (data.size < 4) return false
        val suffix = getIntBigEndian(data, data.size - 4)
        return suffix == Z_SYNC_FLUSH
    }

    private fun reallocate(original: ByteBuffer?, length: Int): ByteBuffer {
        val buffer = ByteBuffer.allocate(length)
        buffer.put(original)
        return buffer
    }

    private fun buffer(data: ByteArray) {
        val notNullFlushBuffer = flushBuffer ?: {
            ByteBuffer.allocate(data.size * 2).apply {
                flushBuffer = this
            }
        }.invoke()
        if (notNullFlushBuffer.capacity() < data.size + notNullFlushBuffer.position()) {
            notNullFlushBuffer.flip()
            flushBuffer = reallocate(
                notNullFlushBuffer, (notNullFlushBuffer.capacity() + data.size) * 2
            )
        }
        notNullFlushBuffer.put(data)
    }

    fun reset() {
        inflater.reset()
    }

    @Throws(DataFormatException::class)
    fun decompress(data: ByteArray): String? {
        var dataAsMutable = data
        if (!isFlush(dataAsMutable)) {
            buffer(dataAsMutable)
            return null
        } else {
            flushBuffer?.let { flushBuffer ->
                buffer(dataAsMutable)
                val arr = flushBuffer.array()
                dataAsMutable = ByteArray(flushBuffer.position())
                System.arraycopy(arr, 0, dataAsMutable, 0, dataAsMutable.size)
                ZlibDecompressor.flushBuffer = null
            }
        }
        val buffer = getDecompressBuffer()
        try {
            InflaterOutputStream(buffer, inflater).use { decompressor ->
                decompressor.write(dataAsMutable)
                return buffer.toString("UTF-8")
            }
        } catch (ex: IOException) {
            throw DataFormatException("Malformed").initCause(ex) as DataFormatException
        } finally {
            if (maxBufferSize < buffer.size()) {
                decompressBuffer = newDecompressBuffer()
            } else {
                buffer.reset()
            }
        }
    }
}